src/main/java/com/ifactory/press/db/solr/search/SafariBlockJoinQuery.java

```diff
@@ -5,339 +5,365 @@
+
+                //System.out.println("  return parentDoc=" + parentDoc + " childDocUpto=" + childDocUpto);
+                return maxScoringDoc;
+            }
+        }
 
-        //System.out.println("  nextChildDoc=" + nextChildDoc);
-        if ((acceptDocs != null && !acceptDocs.get(parentDoc))
-            // shouldn't happen, but it did.  I'm not sure if this is a consequence of our allowing 
-            // parents to be a child -- I don't think so -- it seems more likely the index can just get in 
-            // a state where there are children with no parent, and that could cause this?
-            || parentDoc == -1
-            ) {
-          // Parent doc not accepted; skip child docs until
-          // we hit a new parent doc:
-          do {
-            nextChildDoc = childScorer.nextDoc();
-          } while (nextChildDoc <= parentDoc);
+        @Override
+        public int docID() {
+            return maxScoringDoc;
+        }
 
-          continue;
+        @Override
+        public float score() throws IOException {
+            return maxScore;
         }
 
-        maxScore = Float.NEGATIVE_INFINITY;
-        totalFreq = 0;
-        do {
-          final int childFreq = childScorer.freq();
-          final float childScore = childScorer.score();
-          if (childScore > maxScore) {
-            maxScore = childScore;
-            maxScoringDoc = nextChildDoc;
-          }
-          totalFreq += childFreq;
-          nextChildDoc = childScorer.nextDoc();
-        } while (nextChildDoc <= parentDoc);
-
-        //System.out.println("  return parentDoc=" + parentDoc + " childDocUpto=" + childDocUpto);
-        return maxScoringDoc;
-      }
-    }
+        @Override
+        public int freq() {
+            return totalFreq;
+        }
 
-    @Override
-    public int docID() {
-      return maxScoringDoc;
-    }
+        // @Override  rfhi took out override, is this needed?  This may not be called if some other method is called instead
+        public int advance(int parentTarget) throws IOException {
+
+            //System.out.println("Q.advance parentTarget=" + parentTarget);
+            if (parentTarget == DocIdSetIterator.NO_MORE_DOCS) {
+                return parentDoc = DocIdSetIterator.NO_MORE_DOCS;
+            }
+
+            if (parentTarget == 0) {
+                // Callers should only be passing in a docID from
+                // the parent space, so this means this parent
+                // has no children (it got docID 0), so it cannot
+                // possibly match.  We must handle this case
+                // separately otherwise we pass invalid -1 to
+                // prevSetBit below:
+                return nextDoc();
+            }
+
+            prevParentDoc = parentBits.prevSetBit(parentTarget - 1);
+
+            //System.out.println("  rolled back to prevParentDoc=" + prevParentDoc + " vs parentDoc=" + parentDoc);
+            assert prevParentDoc >= parentDoc;
```

Comment:
One thing that has bitten the test suite is asserts, IIRC the og test suite would have one fail because it had different settings than the actual jar that was produced. This error happened when trying to test the sad path that hit the assert.